File: src/main/java/org/funivan/lologa/algo/ga/genome/population/crossing/RandomCross.java
package org.funivan.lologa.algo.ga.genome.population.crossing;

import org.funivan.lologa.algo.ga.genome.GenomeInterface;

import java.util.HashMap;
import java.util.Random;

public class RandomCross implements CrossingInterface {
    private final Random rand;

    public RandomCross() {
        this.rand = new Random();
    }

    @Override
    public HashMap<String, Double> cross(GenomeInterface father, GenomeInterface mother) {
        HashMap<String, Double> result = new HashMap<>();
        for (final String type : father.data().keySet()) {
            final GenomeInterface target = (this.rand.nextInt(100) > 50) ? father : mother;
            result.put(type, target.data().get(type) );
        }
        return result;
    }


}

File: src/main/java/org/funivan/lologa/algo/ga/genome/population/crossing/RandomCross.kt
package org.funivan.lologa.algo.ga.genome.population.crossing

import org.funivan.lologa.algo.ga.genome.GenomeInterface
import java.util.*

class RandomCross : CrossingInterface {
    private val rand: Random = Random()

    override fun cross(father: GenomeInterface, mother: GenomeInterface): HashMap<String, Double> {
        val result = HashMap<String, Double>()
        for (type in father.data().keys) {
            val target = if (this.rand.nextInt(100) > 50) father else mother
            val value = target.data()[type]
            if (value != null) {
                result.put(type, value)
            }
        }
        return result
    }


}
